spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1270,6 +1273,7 @@ protected void pollAndInvoke() {
 				return;
 			}
 			this.polling.set(true);
+			beforePoll();
```

Comment:
`pollAndInvoke()` starts with offset related operation e.g. it can commit offsets. Are these offsets of the previous record/batch processing? Perhaps it would be better to call `finishInvoke()` just before `idleBetweenPollIfNecessary()` few lines above if single batch of records is fully processed in two `pollAndInvoke()` invocations?

Reply:
I concur that `beforePoll()` should be imediately before the poll. It's not clear why you have 2 calls to `beforePoll()` (one here and on in each branch of `doPoll()`.

However, we now have a new method `pollConsumer()` (common to both batch and record listeners - the `beforePoll()` should be done there.

I don't think that calling `afterRecordsProcessed()` before `processCommits()` will cause any problems; also, moving it from the `finally` block might cause us to miss the last call (when `isRunning()` goes false, when the container is stopped.

Reply:
Two calls to `beforePoll()` method was bug. I removed redundant call when writing `InOrder` tests.

Reply:
Thanks - I will take another look tomorrow; thanks again.